src/DeploymentTemplate.ts

```diff
@@ -395,17 +398,79 @@ export class DeploymentTemplate extends DeploymentDocument {
         const spanOfValueInsideString = tleValue.getSpan();
         return this.getDocumentText(spanOfValueInsideString, parentStringToken.span.startIndex);
     }
+
+    public getCodeLenses(hasAssociatedParameters: boolean): ResolvableCodeLens[] {
+        return this.getParameterCodeLenses(hasAssociatedParameters)
+            .concat(this.getNestedTemplateCodeLenses());
+    }
+
+    private getParameterCodeLenses(hasAssociatedParameters: boolean): ResolvableCodeLens[] {
+        const lenses: ResolvableCodeLens[] = [];
+
+        // Code lens for the "parameters" section itself - indicates currently-selected parameter file and allows
+        // user to chnage it
+        const parametersCodeLensSpan = this.topLevelValue?.getProperty(templateKeys.parameters)?.span
+            ?? new language.Span(0, 0);
+        if (hasAssociatedParameters) {
+            lenses.push(new ShowCurrentParameterFileCodeLens(this, parametersCodeLensSpan));
+        }
+        lenses.push(new SelectParameterFileCodeLens(this, parametersCodeLensSpan));
+
+        if (hasAssociatedParameters) {
+            // Code lens for each parameter definition
+            lenses.push(...this.topLevelScope.parameterDefinitions.map(pd => new ParameterDefinitionCodeLens(this, pd)));
+        }
+
+        return lenses;
+    }
+
+    private getNestedTemplateCodeLenses(): ResolvableCodeLens[] {
+        const lenses: ResolvableCodeLens[] = [];
+        const allScopes = this.findAllScopes();
+        for (let scope of allScopes) {
+            if (scope.rootObject) {
+                const lens = NestedTemplateCodeLen.create(this, scope.rootObject.span, scope.scopeKind);
+                if (lens) {
+                    lenses.push(lens);
+                }
+            }
+        }
+
+        return lenses;
+    }
+
+    public findAllScopes(): TemplateScope[] {
+        const allScopes: TemplateScope[] = [];
+        traverse(this.topLevelScope);
+        return allScopes;
+
+        function traverse(scope: TemplateScope | undefined): void {
+            for (let childScope of scope?.childScopes ?? []) {
+                if (allScopes.indexOf(childScope) < 0) {
+                    allScopes.push(childScope);
+                } else {
+                    let a = 1;
+                    a = a;
+                }
```

Comment:
Why are we doing this?

Reply:
I know got rid of that in some branch somewhere...  :-)

It was just a place to put a breakpoint for easy debugging.